aten/src/ATen/core/Generator.h

```diff
@@ -1,24 +1,62 @@
 #pragma once
 
-#include <ATen/core/ATenGeneral.h>
 #include <stdint.h>
+#include <memory>
+#include <mutex>
+#include <deque>
+#include <atomic>
+#include <typeinfo>
+#include <utility>
+
+#include <c10/util/Exception.h>
+#include <c10/util/C++17.h>
+#include <ATen/Device.h>
+
+/*
+* Generator note.
+* A Pseudo Random Number Generator (PRNG) is an engine that uses an algorithm to 
+* generate a seemingly random sequence of numbers, that may be later be used in creating 
+* a random distribution. Such an engine almost always maintains a state and requires a
+* seed to start off the creation of random numbers. Often times, users have
+* encountered that it could be beneficial to be able to create, retain, and destroy 
+* PRNG states and also be able to have control over the seed value.
+*
+* A Generator in ATen gives users the ability to read, write and modify a PRNG engine.
+* For instance, it does so by letting users seed a PRNG engine, fork the state of the
+* engine, etc.
+*
+* By default, there is one generator per device, and a device's generator is 
+* lazily created. A user can use the torch.Generator() api to create their own generator.
+* Currently torch.Generator() can only create a CPUGenerator.
+*/
 
 namespace at {
 
+constexpr uint64_t default_rng_seed_val = 67280421310721;
+
 struct CAFFE2_API Generator {
-  Generator() {};
-  Generator(const Generator& other) = delete;
-  Generator(Generator&& other) = delete;
-  virtual ~Generator() {};
-
-  virtual Generator& copy(const Generator& other) = 0;
-  virtual Generator& free() = 0;
-
-  virtual uint64_t seed() = 0;
-  virtual uint64_t initialSeed() = 0;
-  virtual Generator& manualSeed(uint64_t seed) = 0;
-  virtual Generator& manualSeedAll(uint64_t seed) = 0;
-  virtual void * unsafeGetTH() = 0;
+  // Constructors
+  Generator(Device device_in, uint64_t seed_in);
+  Generator(const Generator& other);
```

Comment:
This copy constructor seems pretty dangerous, because you're planning to subclass this class, and add fields to the subclasses. When you allow copy construction of classes with subclasses, you are exposing yourself to the possibility of object slicing (https://stackoverflow.com/questions/274626/what-is-object-slicing). It is probably best if we don't allow copy constructor at all, and instead have a virtual method for doing clones.

Reply:
So, we should disable all copy/move constructors on the superclass Generator.

Reply:
it's not an object slicing issue as long as you take `other` per reference, but disabling copy constructors where not needed is always a good idea.